mitxgraders/formulagrader.py

```diff
@@ -437,11 +446,58 @@ def check_response(self, answer, student_input):
                 msg = "Invalid Input: Could not parse '{}' as a formula"
                 raise InvalidInput(msg.format(student_input))
 
+    def generate_variable_list(self, answer, student_input):
+        """
+        Generates the list of variables required to perform a comparison and the
+        corresponding sampling dictionary, taking into account any numbered variables.
+
+        Returns variable_list, sample_from_dict
+        """
+        # Pre-parse all expressions (these all get cached)
+        parsers = [
+            parsercache.get_parser(expr, self.config['case_sensitive'], self.suffixes)
+            for expr in answer['expect']['comparer_params']
+            ]
+        # If the student input is not empty, parse that too
+        if not (student_input is None or student_input.strip() == ""):
+            parsers.append(parsercache.get_parser(student_input,
+                                                  self.config['case_sensitive'],
+                                                  self.suffixes))
+        # Create a list of all variables used in the expressions
+        vars_used = set().union(*[parser.variables_used for parser in parsers])
+
+        # Initiate the variables list and sample_from dictionary
+        variable_list = self.config['variables'][:]
+        sample_from_dict = self.config['sample_from'].copy()
+
+        # Find all unassigned variables
+        cs = self.config['case_sensitive']
+        bad_vars = set(var for var in vars_used if casify(var, cs) not in variable_list)
+
+        # Check to see if any unassigned variables are numbered_vars
+        for var in bad_vars:
+            csvar = casify(var, cs)
```

Comment:
Looking through this now. I got stuck on the next few lines for quite a while :). 

I think it's worth abstracting the next several lines into a separate function for (1) clarity and (2) testing that isolated bit of logic. Maybe something like:

```python
for var in bad_vars:
    if numbered_variables.match(var):
        # do stuff
```

Also, I think it will be more understandable to just construct a single regexp for matching the whole numbered_variable. With the help of SO, I made this function that I think does what we need (sans casification, but that either shouldn't be hard or we'll drop it.)

```python
import re

def numbered_vars_regexp(variable_fronts):
    """
    Creates a regexp to match numbered variables.

    Arguments:
        variable_fronts ([str]): a list of variable fronts

    Usage
    =====

    Matches numbered variables:
    >>> regexp = numbered_vars_regexp(['b', 'c', 'cat'])
    >>> regexp.match('b_{12}').groups()
    ('b_{12}',)
    >>> regexp.match('b_{-3}').groups()
    ('b_{-3}',)
    >>> regexp.match('b_{0}').groups()
    ('b_{0}',)

    Other variables match, too:
    >>> regexp.match('cat_{17}').groups()
    ('cat_{17}',)

    Stuff that shouldn't match does not match:
    >>> regexp.match('b') == None
    True
    >>> regexp.match('b_{05}') == None
    True
    >>> regexp.match('b_{-05}') == None
    True
    """
    # Adapted from https://stackoverflow.com/a/33411504/2747370

    # I'm not sure that the sorting is really necessary, since we match underscore next
    longest_first = sorted(variable_fronts, key=len, reverse=False)
    base_regexp = (r"^((?:{longest_first})" # match base
                   "_{{" # match _{, double curly b/c of format string
                   "(?:[-]?[1-9]\d*|0)" # match number pattern
                   "}})$") # match closing }
    return re.compile(base_regexp.format(
        longest_first = '|'.join(map(re.escape, longest_first))
    ))

```

Reply:
Yeah, I realized I could do the whole thing with regex, but knew how to strip off some of the simpler bits without getting into the mud of regex ;-) I'll go and make that change now.

Reply:
> without getting into the mud of regex ;-) 

Yes, I was a bit wary of wading into that myself. I heard a story about a guy who had a problem, and tried to solve it with a regex, and then he had two problems. (That's supposed to be a joke.) But I think it's worth it here, especially for making an easier-to-test bit of logic.